tests/lib/k8s/caasp.py

```diff
@@ -0,0 +1,135 @@
+        self.SSH_AGENT_PID = res.stdout.decode(
+            'utf-8').split(';')[2].split('=')[1]
+        os.environ['SSH_AUTH_SOCK'] = self.SSH_AUTH_SOCK
+        os.environ['SSH_AGENT_PID'] = self.SSH_AGENT_PID
+        try:
+            res = subprocess.run(['ssh-add', self.hardware._private_key],
+                                 check=True)
+        except subprocess.CalledProcessError:
+            msg = 'Failed to add keys to agent'
+            logger.error(msg)
+            raise Exception(msg)
+
+    def destroy(self, skip=False):
+        logger.info(f"kube destroy on hardware {self.hardware}")
+        if skip:
+            # We can skip in most cases since the nodes themselves will be
+            # destroyed instead.
+            return
+        # This kills the SSH_AGENT_PID agent
+        try:
+            subprocess.run(['ssh-agent', '-k'], check=True)
+        except subprocess.CalledProcessError:
+            logger.exception(f'Killing ssh-agent with PID \
+{self.SSH_AGENT_PID} failed')
+
+        # TODO(jhesketh): Uninstall kubernetes
+
+    def install_kubernetes(self):
+        super().install_kubernetes()
+        self.hardware.execute_ansible_play_raw('playbook_caasp.yaml')
+        self.hardware.get_masters()
+        self.hardware.get_workers()
+        self._caasp_init()
+        with cd(self.clusterpath):
+            self._caasp_bootstrap()
+        with cd(self.clusterpath):
+            self._caasp_join()
+
+    def _caasp_init(self):
+        try:
+            res = subprocess.run(
+                ['skuba', 'cluster', 'init', '--control-plane',
+                 self.hardware.masters[0].get_ssh_ip(),
+                 self.clusterpath], check=True)
+            logger.debug(res.args)
+        except subprocess.CalledProcessError:
+            msg = 'Cluster init step failed'
+            logger.error(msg)
+            raise Exception(msg)
+
+    def _caasp_bootstrap(self):
+        try:
+            res = subprocess.run(
+                ['skuba', 'node', 'bootstrap', '--user', 'sles', '--sudo',
+                 '--target', self.hardware.masters[0].get_ssh_ip(),
+                 self.hardware.masters[0].dnsname], check=True)
+            logger.debug(res.args)
+        except subprocess.CalledProcessError:
+            msg = 'Cluster bootsrap step failed'
+            logger.error(msg)
+            raise Exception(msg)
+
+    def _caasp_join(self):
+        for worker in self.hardware.workers:
+            try:
+                res = subprocess.run(
+                    ['skuba', 'node', 'join', '--role', 'worker',
+                     '--user', 'sles', '--sudo', '--target',
+                     worker.get_ssh_ip(), worker.dnsname], check=True)
+                logger.debug(res.args)
+            except subprocess.CalledProcessError:
+                msg = f'Node {worker.dnsname} failed to join cluster'
+                logger.error(msg)
+                raise Exception(msg)
+
+
+class cd:
```

Comment:
Using `@contextlib.contextmanager` with `try: finally` would be imo better to make sure we go back even in an error case.

Reply:
Done